week-1-ts-ecosystem/w01e03-object-oriented-structure/src/models/product/ProductWithPrice.ts

```diff
@@ -0,0 +1,37 @@
+import { Product } from './Product';
+
+abstract class ProductWithPrice extends Product {
+  private _price = 0;
+
+  public constructor({
+    price,
+    ...product
+  }: Pick<ProductWithPrice, 'amount' | 'name' | 'price'>) {
+    super(product as Product);
+    this.price = price;
+  }
+
+  public set price(value: number) {
+    if (value > 0) {
+      this._price = value;
+    } else {
+      throw Error('Invalid price');
+    }
+  }
+
+  public getPrice(): number {
+    return this._price;
+  }
+
+  public update({
+    price,
+    ...product
+  }: Partial<Pick<ProductWithPrice, 'amount' | 'name' | 'price'>>): void {
```

Comment:
Jak powtarzają Ci się zapisy `Partial<Pick<`

To możesz zorbić sobie własny `Util` type:

```typescript
type Sliced<T, K extends keyof T> = Partial<Pick<T, K>>
```

```typescript
Sliced<ProductWithPrice, 'amount' | 'name' | 'price'>
```

nieco wygodniejszy zapis. (nazwa jest przykładowa)